java/core/src/main/java/com/google/protobuf/UnsafeUtil.java

```diff
@@ -259,6 +260,26 @@ static long addressOffset(ByteBuffer buffer) {
     return MEMORY_ACCESSOR.getLong(buffer, BUFFER_ADDRESS_OFFSET);
   }
 
+  /**
+   * Returns a new {@link String} backed by the given {@code chars}. The char array should not
+   * be mutated any more after calling this function.
+   */
+  static String moveToString(char[] chars) {
+    if (STRING_VALUE_OFFSET == -1) {
+      // In the off-chance that this JDK does not implement String as we'd expect, just do a copy.
+      return new String(chars);
+    }
+    final String str;
+    try {
+      str = (String) UNSAFE.allocateInstance(String.class);
+    } catch (InstantiationException e) {
+      // This should never happen, but return a copy as a fallback just in case.
+      return new String(chars);
+    }
+    putObject(str, STRING_VALUE_OFFSET, chars);
```

Comment:
Copy-paste the comment from Luke:

Seems like this undermines the memory semantics of the value field.

The value field in string is marked final, but this is a non-volatile write.  So if this string is unsafely published (which is possible using the APIs this is exposed through), then people could get NPEs when calling methods on this string.

There is also the other problem about the other fields in string.  In Java7/8 it doesn't have any (besides hashCode), but in the past it has had a length and offset field and in the future it might have other fields.  How do we make sure that this code works on future JDK versions, or JDKs with different field semantics.  Or things like -XX:+UseCompressedStrings in jdk6 which changed the field to be a byte[] instead of a char[]

For example in jdk10 it looks like there is an extra byte field
http://hg.openjdk.java.net/jdk10/master/file/be620a591379/src/java.base/share/classes/java/lang/String.java


To fix the memory ordering problem you can probably just use putObjectVolatile, as for handling compatibility with multiple future jdk versions, im not sure.

Reply:
I thought the memory ordering would work because we both allocate and populate the variable in this same function - it should not be possible to dereference `value` until the object is initialized (this code should be effectively the same as the code in the standard constructor). But perhaps inlining + escape analysis breaks that assumption - are you seeing that? Doing a volatile write would indeed fix that.

As for JDK9+, they don't use a `char[]` for `value` so don't execute this code (guarded by the first if). I figured this would keep it future proof for all practical purposes, and in the fairly inconveivable chance that `char[] value` is restored in a future version a change could be made then. The speedup on JRE8 is reasonable, but not a deal breaker so it would be ok to remove this function too.

Reply:
The memory semantics of final fields and constructors are fairly well documented in JLS
https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.4

the key thing is that final field writes force a memory barrier at the end of the constructor.  But since you aren't using the constructor this doesn't work.  I believe volatile will fix the problem in the current code in JDK8, but it can still cause other problems in the future. 

I think the function should be removed and the string constructor should simply be invoked.  The caller of this function actually performs an Arrays.copyOf to correctly size the array.  instead we could just call the String(char[], int, int) constructor and do both things at the same time.  At least in some cases it would be just as good as the current unsafe code

Reply:
Ah yeah, I was thinking of the function return here as a memory fence for some reason, but obviously that's not correct.

I still think the chance of a problem in future Java versions is extremely low since the method is guarded by the presence of `char[] value`, but at the same time that's enough reason to avoid this  hacky optimization if it only applies to older (Java8) runtimes. Will send a PR